### PR TITLE
LayerNorm

### PR DESCRIPTION
**Pull request by Kaixhin**
WIP. See https://github.com/pytorch/pytorch/pull/2019 for details on this.

### CODE REVIEW

**Comment by Kaixhin** [comment]
Pinging @jekbradbury to read the discussion in the previous PR and do a sanity check on this one.

File: torch/nn/modules/instancenorm.py [hunk 1]
```diff
@@ -36,29 +36,90 @@ def eval(self):
...
+    for each object in a mini-batch. Gamma and beta are learnable parameter
+    vectors of size C (where C is the input size).
+
+    During training, this layer keeps a running estimate of its computed mean
+    and variance. The running sum is kept with a default momentum of 0.1.
+
+    At evaluation time (`.eval()`), the default behaviour of the LayerNorm
+    module stays the same, i.e. the running mean/variance is NOT used for
+    normalization. One can force using the stored mean and variance with
+    the `.train(False)` method.
```
Comment (colesbury) [inline torch/nn/modules/instancenorm.py]: How does this work? As far as I can tell, `eval()` just calls `train(False)`. I see that `_InstanceNorm` overrides `eval()`, but:

1) eval() isn't called recursively. If you call `eval()` on a container, it'll call `train(False)` on it's children.
2) `_InstanceNorm` doesn't appear to use `self.training` so I don't see how you can "force" using mean/var.

**Comment by jekbradbury** [comment]
I'm not entirely sold on reusing the InstanceNorm code to do LayerNorm. I think Sam's comments are right, and additionally I'm afraid that keeping running statistics (which is inconsistent with the motivation of LayerNorm) would be unnecessary overhead. LayerNorm is a surprisingly significant fraction of compute time for Transformer-like networks already; it might be worth doing a microbenchmark to make sure that it's as fast as we can easily make it.

File: torch/nn/modules/instancenorm.py (hunk 1 above)
Comment (Kaixhin) [inline torch/nn/modules/instancenorm.py]: Agreed, that code looked mysterious to me and I wasn't sure it worked either. Pinging @DmitryUlyanov to find out more. As James said below, using running statistics seems to be counter to layer/instance norm, and if so, then we could actually change InstanceNorm to use what I wrote for LayerNorm (I'll keep the instance norm name, just update the docs to refer to layer norm too): https://github.com/pytorch/pytorch/pull/2019

Comment (DmitryUlyanov) [inline torch/nn/modules/instancenorm.py]: Hi, batchnorm has hardcoded `train=True` here https://github.com/Kaixhin/pytorch/blob/c37529686a0be1e6526cea3094a96dad0510aebe/torch/nn/modules/instancenorm.py#L27, so `self.trainining` is not used at all.

**Comment by Kaixhin** [comment]
Layer norm and instance norm are similar normalisation strategies that are batch-independent and therefore should not require running statistics. As noted above, this was not working for `nn.InstanceNorm{1,2,3}d` anyway. Hence, we can move away from using batch norm as a backend, and furthermore, we can make them parameterless by default.

To clarify the difference between these two, layer norm works on 2D inputs, normalising over the 2nd dim, whilst instance norm works on 3/4/5D inputs, normalising over all except for the first 2 dims.

There I propose the following for this PR:

- [x] Replace the batch norm backend for instance norm with a separate function (`F.instance_norm`), which is by default parameterless and hence can work with arbitrary 3+D inputs.
- [x] Implement layer norm using @jekbradbury's simple code from https://github.com/pytorch/pytorch/issues/1959. It seems simpler to do this in the Module itself rather than adapt `F.instance_norm` to work with less than 3 dimensions.
- [x] Add basic module tests for `nn.LayerNorm` and `nn.InstanceNorm{1,2,3}d`.
- [ ] Fix `test_LayerNorm`.
- [ ] Fix `_test_InstanceNorm`.
- [x] Tidy up the normalisation docs to make what happens in layer and instance norm clear.

**Comment by ngimel** [comment]
@jekbradbury, if LayerNorm is becoming a computational bottleneck, may be it makes sense to put it in the backend? Should be pretty similar to batchNorm.

**Comment by ngimel** [comment]
I must be missing something here, but why are we moving from using batch norm as a backend? The overhead in batch norm to keep running statistics should be small (at least as long as batch_size * num_channels >> feature_size), but everything is done in 1 kernel, which should be a big performance advantage over 1 kernel to calculate mean, a handful to calculate std and another handful to normalize. Using cunn backend might be better here than cudnn, since it does not care if batch_size is 1 (which it will be after reshaping), IIRC, cudnn will have noticeably worse perf with batch_size 1.

**Comment by Kaixhin** [comment]
@ngimel What would you suggest, considering [instance norm uses a batch size of 1](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/instancenorm.py#L23) and [`batch_norm`](https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L560-L563) doesn't (sensibly so) provide a way to turn off `cudnn`?

I can also imagine situations where `batch_size` * `num_channels` << `feature_size` - in the lower layers of a CNN where you're looking at activation statistics for style transfer for example. James may also be able to say more about his use cases.

**Comment by ngimel** [comment]
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L560-L563 just links to autograd batchnorm, and there it's not too hard to not use cudnn if batch size == 1: https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/functions/batch_normalization.cpp#L55-L58. 
Oops, had << the wrong way, if batch_size*num_channels << feature_size, overhead is small, the bigger num_channels and smaller feature_size, the bigger the overhead is (because the size of the running_vars would be batch_size * num_channels), so it is the upper layers of cnn that would suffer. Idk, the effort one might want to spend on it depends on whether it is a bottleneck or not, if it is, then modifying cunn's backend not to track running vars might be worth it.

**Comment by Kaixhin** [comment]
Note for when coming back to this - may want to make LayerNorm the same for {1, 2, 3}D for the same reasons as underlined in https://github.com/pytorch/pytorch/issues/2628.

**Comment by netheril96** [comment]
I cursively read the comments, but still don't understand why two separate PRs.

**Comment by Kaixhin** [comment]
Closing in favour of https://github.com/pytorch/pytorch/pull/4922.
